src/storage/engine/mod.rs

```diff
@@ -33,7 +33,17 @@ pub const TEMP_DIR: &'static str = "";
 const SEEK_BOUND: usize = 30;
 const DEFAULT_TIMEOUT_SECS: u64 = 5;
 
-pub type Callback<T> = Box<FnBox(Result<T>) + Send>;
+pub type Callback<T> = Box<FnBox((CbContext, Result<T>)) + Send>;
+
+pub struct CbContext {
```

Comment:
First of all, old API does not support underlying engine to return more data other than `Result`, we need to change it no matter we reuse `Context` or not.
In addition, the content of the two Context are different, and subsequent changes will make them more different. I see no reason to reuse it.